feat: create new blog post (#2838)
agrc/gis.utah.gov: src/content/blog/2024-08-22-the-ugrc-api-address-location-lane.mdx

---
title: The UGRC API address location lane
author: Steve Gourley
description: >-
  The UGRC API is a mature service that provides address locating. We are very proud to offer this service for free, but it's not the only solution for every addressing problem. Read this to understand the limitations and capabilities of our geocoding endpoint.
date: '2024-08-22T23:47:42.633Z'
category: API
cover_image: /src/images/pillar-blog/default-social-card.png
cover_image_alt: ugrc social card
---

import centerlineImage from '@images/blog/2024-08-22-the-ugrc-api-address-location-lane/centerline-address-match-example.png';
import { Image } from 'astro:assets';

> The [UGRC API](https://api.mapserv.utah.gov/) is a valuable resource for [address locating](https://api.mapserv.utah.gov/docs/v1/endpoints/geocoding/). We are proud to offer this service to you for free, however it's not the only solution for your addressing questions. Let's explore the layers that enable the address locating `/geocode` endpoint to help you understand what this service offers and its limitations.

Reviewer: ```suggestion
The [UGRC API](https://api.mapserv.utah.gov/) is a valuable resource for [locating an address](https://api.mapserv.utah.gov/docs/v1/endpoints/geocoding/). We are proud to offer this service to you for free, however, it's not the only solution for your address locating. Let's explore the layers that enable the `/geocode` endpoint to help you understand what this service offers and its limitations.
```

The road centerlines and address point data is regularly updated, with the "Big 5" counties (Salt Lake, Utah, Davis, Weber, Washington) providing monthly updates. While the other counties provide updates on a less frequent schedule, the "Big 5" cover approximately 80% of Utah's population, ensuring great coverage in densely populated areas. Stay informed about the latest updates by following the [SGID Update](/blog/categories/sgid-updates/) blog category.

## The data limitations

> While the UGRC API address locating endpoint is a powerful service, it's important to be aware of its limitations. The accuracy of our data depends on the quality and timeliness of the data provided by our address authority partners. In rural counties, where resources and funding may be limited, data gaps exist. This is most pronounced in counties like Juab, Piute, Uintah, and Sevier. We are working hard to improve the data quality, especially in these areas, but it's a work in progress. These data characteristics make it problematic to solely use our API for address validation for mail delivery, 911 response, or voter registration types of questions.

Reviewer: @ZachBeck and @MichaelAGRC could you confirm that those are the rural counties where most of our data gaps exist?

Reviewer 2: I'd add Daggett, Duchesne and Rich to that list.

Reviewer 3: Wow that's like 25 percent of the counties but probably a very small population of addresses.

Reviewer 4: Sevier and Piute shouldn't be on the list. It should be Daggett, Duchesne, Juab, Rich, Sanpete, Uintah, and maybe San Juan

> While the UGRC API address locating endpoint is a powerful service, it's important to be aware of its limitations. The accuracy of our data depends on the quality and timeliness of the data provided by our address authority partners. In rural counties, where resources and funding may be limited, data gaps exist. This is most pronounced in counties like Daggett, Duchesne, Juab, Piute, Uintah, Rich, and Sevier. We are working hard to improve the data quality, especially in these areas, but it's a work in progress. These data characteristics make it problematic to solely use our API for address validation for mail delivery, 911 response, or voter registration types of questions.

Reviewer: @MichaelAGRC can you confirm this list of counties where data gaps exist? does that jive with your data requests and road processing?

Reviewer 2: Looks like you incorporated the counties I mentioned, so yes.

> While the UGRC API address locating endpoint is a powerful service, it's important to be aware of its limitations. The accuracy of our data depends on the quality and timeliness of the data provided by our address authority partners. In rural counties, where resources and funding may be limited, data gaps exist. This is most pronounced in counties like Daggett, Duchesne, Juab, Piute, Uintah, Rich, and Sevier. We are working hard to improve the data quality, especially in these areas, but it's a work in progress. These data characteristics make it problematic to solely use our API for address validation for mail delivery, 911 response, or voter registration types of questions.

Reviewer 3: Is this too on the nose?

> While the UGRC API address locating endpoint is a powerful service, it's important to be aware of its limitations. The accuracy of our data depends on the quality and timeliness of the data provided by our address authority partners. In rural counties, where resources and funding may be limited, data gaps exist. This is most pronounced in counties like Daggett, Duchesne, Juab, Piute, Uintah, Rich, and Sevier. We are working hard to improve the data quality, especially in these areas, but it's a work in progress. These data characteristics make it problematic to solely use our API for address validation for mail delivery, 911 response, or voter registration types of questions.

Reviewer 3: I agree about adding Sanpete since we just had that issue with their data with the school board. 

Greg suggested these in the first place and Michael agreed. Is there a difference between your address point opinion and their road opinion?

## Road centerlines versus address points

> Road centerlines and address points are two different datasets that work together to provide the best address location match rates. By default, both datasets are used to provide the solution and address points have priority in the case of a tie. The UGRC API provides which dataset solved your request in the response and you are able to specify which data participates in the match using the `locator` optional input. This is important since the road centerlines are used to determine the location of an address along a road segment, while the address points are used to determine the exact location of an address. By using both datasets together, you will get the best match rate our data can provide. But a match on the road centerlines has a different meaning than a match on the address points. But if you are zooming, routing, extracting data, or looking for a generalized location for analysis a match on the road centerlines is a great option.

Reviewer: do you want to say anything along the lines that we strive to find a match in the address points first? aka: if we get 100% match in both datasets, we go with the address point.

Reviewer 3: > and address points have priority in the case of a tie.

Reviewer: good point. nice.

<Image
  src={centerlineImage}
  width={800}
  alt="A line indicating fiction street with a circle at the 25% location along the segment"
/>

> To understand why, lets break down what happens with a match on the road centerlines with an example. Consider the input street address is `125 South Fiction Street`. The road centerline data has a segment representing Fiction Street with attributes defining the address range for that segment e.g. 100-200. This guarantees that any address on this segment will have numbers between 100 and 200. Since 125 falls within the address range of the segment, the road centerline address locator will provide a match that is 25% along the line segment in the positive direction. However, the exact location of the address along the segment is not known; It may not exist in terms of a structure, mail delivery, or the availability of 911 services. **If you are solving or validating for mail delivery, 911 response, voter registration, etc., this is not an accurate solution.** You may need to evaluate services provided by the [USPS](https://www.usps.com/) or [Smarty](https://smarty.com/) to validate mail delivery.

Reviewer: ```suggestion
To understand why, lets break down what happens with a match on the road centerlines with an example. Consider the input street address is `125 South Fiction Street`. The road centerline data has a segment representing Fiction Street with attributes defining the address range for that segment e.g. 100-200. This guarantees that any address on this segment will have numbers between 100 and 200. Since 125 falls within the address range of the segment, the road centerline address locator will provide a match that is 25% along the line segment in the positive direction. However, the exact location of the address along the segment is not known; It may not exist in terms of a structure, mail delivery, or the availability of 911 services. **If you are solving or validating for mail delivery, this is not an accurate solution and you may need to evaluate services provided by the [USPS](https://www.usps.com/) or [Smarty](https://smarty.com/).**
```

> Address location matches solved by the address points do represent a geographic location that has been assigned a US Postal Service (USPS) address by the local address authority (i.e., county or municipality) but does not necessarily receive mail. For example, mail may be delivered to a post office box.

Reviewer: ```suggestion
Address location matches solved by the address points represent a geographic location that has been assigned a civic address by the local address authority (i.e., county or municipality) but does not necessarily receive mail. For example, mail may be delivered to a post office box.
```

Reviewer 3: This was taken directly from our address points product page. Do you want to update that to match these suggestions?

Reviewer: @ZachBeck what are your thoughts?

Reviewer 4: I think this makes more sense than mentioning the USPS like on the address page.

Reviewer 3: @gregbunce or @ZachBeck do you mind synchronizing these changes with the address product page and make sure it's also reflected in @miriamseely's metadata work?

## The UGRC benefit

The UGRC API address locating endpoint is a free service that provides the most comprehensive government-sourced GIS address data available for Utah. We are able to communicate with our address authorities to fix any issues you may encounter with the data quickly. This feedback loop ensures that the data is always improving and that you are getting the best address locating service possible. We provide excellent customer service and timely responses to your questions and concerns. We are here to help you get the most out of our data.

> Please carefully evaluate if the UGRC API is the right choice for your addressing needs. If address verification and validation are terms that come to mind, distill what that means to you and your needs. If you are looking for a generalized location for analysis, the UGRC API is a great choice. If you are looking for a precise location for mail delivery, 911 response, voter registration, etc., you may need to evaluate other services.

Reviewer: ```suggestion
Please carefully evaluate if the UGRC API is the right choice for your addressing needs. If you are looking for a generalized location for analysis, the UGRC API is a great choice. If you are looking for a precise location for mail delivery you may need to evaluate other services.
```

Reviewer: i struggle with saying that our data is not suitable for 911 and elections, b/c it is used, in a sense, in both of those systems. 

from my understanding, our address data supplements the voter registration data and is the main input for 911 call routing. 

maybe it's good enough to just say that we know our data is not perfect, but are working with the local addressing authorities to improve it and that 911 and elections are the main drivers that help us improve the data.

Reviewer 3: Right I hear that. But that "in a sense" caveat should be explained and is the intention of this blog post. Can you elaborate on that and update your suggestion? Maybe we add 2 more headings for voting and 911 to explain how they are used?

Reviewer: Sure, I like that. The main takeaway should be that our data is used in conjunction with other data sources in these critical business systems. 

For 911:
we could mention that 911 uses our address data (address points and road centerlines) to route 911 calls originating from a landline (with an attached home address). We work with the service providers to ensure that home addresses for their landline customers are in our system.  
Additionally, in dispatch centers, our address data is used to log incident reports and to route emergency response, however, some dispatch centers have supplemental address data that is used in conjunction with our data. Their supplemental data often has a local nuance, to account for their local vernacular. 

For elections:
The UGRC address locator is an input into the voter registration database. Our data is used to locate voters. However, our data is supplemented with Smarty as well as local knowledge/data: clerks use UGRC address data as a starting point, but can modify the x,y locations in their local copy as needed).  Smarty also provides information regarding residential or commercial)

Reviewer: note: i modified my original suggestion.

Reviewer 3: i took what you added and reorganized. I'm going to resolve this and set the new draft as ready for review again.
